packages/mattermost-redux/src/reducers/entities/threads.ts

```diff
@@ -243,17 +295,84 @@ export const countsReducer = (state: ThreadsState['counts'] = {}, action: Generi
     }
 
     case UserTypes.LOGOUT_SUCCESS:
+        return {};
+    case ChannelTypes.LEAVE_CHANNEL: {
+        if (!extra.threadsToDelete || extra.threadsToDelete.length === 0) {
+            return state;
+        }
+
+        const teamId = action.data.team_id;
+
+        const {unreadMentions, unreadThreads} = extra.threadsToDelete.reduce((curr, item: UserThread) => {
+            curr.unreadMentions += item.unread_mentions;
+            curr.unreadThreads = item.unread_replies > 0 ? curr.unreadThreads + 1 : curr.unreadThreads;
+            return curr;
+        }, {unreadMentions: 0, unreadThreads: 0});
+
+        const {total, total_unread_mentions: totalUnreadMentions, total_unread_threads: totalUnreadThreads} = state[teamId];
+
         return {
-            total: 0,
-            total_unread_threads: 0,
-            total_unread_mentions: 0,
+            ...state,
+            [teamId]: {
+                total: Math.max(total - extra.threadsToDelete.length, 0),
+                total_unread_mentions: Math.max(totalUnreadMentions - unreadMentions, 0),
+                total_unread_threads: Math.max(totalUnreadThreads - unreadThreads, 0),
+            },
         };
     }
+    }
     return state;
 };
 
-export default combineReducers({
-    threads: threadsReducer,
-    threadsInTeam: threadsInTeamReducer,
-    counts: countsReducer,
-});
+function getThreadsOfChannel(threads: ThreadsState['threads'], channelId: string) {
+    const threadsToDelete: UserThread[] = [];
+    for (const rootId of Object.keys(threads)) {
+        if (
+            threads[rootId] &&
+            threads[rootId].post &&
+            threads[rootId].post.channel_id === channelId
+        ) {
+            threadsToDelete.push(threads[rootId]);
+        }
+    }
+
+    return threadsToDelete;
+}
+
+// custom combineReducers function
+// enables passing data between reducers
+function reducer(state: ThreadsState = {threads: {}, threadsInTeam: {}, counts: {}}, action: GenericAction): ThreadsState {
+    const extra: ExtraData = {};
+
+    // acting as a 'middleware'
+    if (action.type === ChannelTypes.LEAVE_CHANNEL) {
+        if (!action.data.viewArchivedChannels) {
+            extra.threadsToDelete = getThreadsOfChannel(state.threads, action.data.id);
+        }
+    }
```

Comment:
Not sure if we've done this type of thing before. Looks good to me but interested to see what @hmhealey thinks

Reply:
@jwilander we've done it in the posts reducer, well not quite the same but same pattern.
my opinion is that It's not something we should do often but I couldn't figure how to update counts and threads in team without having that extra data.

Reply:
I wish we had a better pattern for doing this, but yeah, it's acceptable to do this. There's part of the redux docs that talks about how combineReducer is convenient for some things but limited sometimes.

Reply:
Another way I can think of achieving just the same, is a redux middleware that adds extra data just on this action, which is pretty close to this, but feels like an overkill